p2p/src/net/libp2p/mod.rs

```diff
@@ -101,11 +105,41 @@ fn parse_discovered_addr(peer_id: PeerId, peer_addr: Multiaddr) -> Option<Multia
     }
 }
 
+/// Parse all discovered addresses and group them by PeerId
+fn parse_peers<T>(peers: Vec<(PeerId, Multiaddr)>) -> Vec<net::AddrInfo<T>>
```

Comment:
Whenever I see `mut`, I feel allergic.

So this function seems to be doing the same thing for every peer, then putting the result in a Vec. Well, why isn't this isolated per peer, and then an `into_iter()` of the Vec is used to consume it to the desired result? Unless I'm missing something, this can be made easier on the eyes.

Since peers are independent, please write a function that processes each peer individually, then, iter() over these function calls and collect the result.

Reply:
We get a vector of `(PeerId, MultiAddr)` tuples from libp2p and there is not other grouping so first we need to parse each address to verify that it's something Mintlayer supports and then group all the valid addresses under the given `PeerId`.

So basically this list (which is a valid list of discovered addresses):
```rust
vec![
    ("12D3KooWRn14SemPVxwzdQNg8e8Trythiww1FWrNfPbukYBmZEbJ", "/ip6/::1/tcp/9091"),
    ("12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c", "/ip6/::1/tcp/5555"),
    ("12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c", "/dns4/foo.com/tcp/80/http"),
    ("12D3KooWGK4RzvNeioS9aXdzmYXU3mgDrRPjQd8SVyXCkHNxLbWN", "/ip4/127.0.0.1/tcp/80/http"),
    ("12D3KooWRn14SemPVxwzdQNg8e8Trythiww1FWrNfPbukYBmZEbJ", "/dns4/foo.com/tcp/80/http"),
    ("12D3KooWGK4RzvNeioS9aXdzmYXU3mgDrRPjQd8SVyXCkHNxLbWN", "/ip4/127.0.0.1/tcp/8080"),
    ("12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c", "/ip4/127.0.0.1/tcp/6666/quic"),
    ("12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c", "/ip4/127.0.0.1/tcp/9090/quic"),
    ("12D3KooWRn14SemPVxwzdQNg8e8Trythiww1FWrNfPbukYBmZEbJ", "/ip4/127.0.0.1/tcp/7777"),
    ("12D3KooWGK4RzvNeioS9aXdzmYXU3mgDrRPjQd8SVyXCkHNxLbWN", "/ip6/::1/tcp/8081"),
    ("12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c", "/dns6/foo.com/tcp/443/https"),
]
```
is converted into an address information list, like this:
```rust
vec![
    AddrInfo {
        id: "12D3KooWRn14SemPVxwzdQNg8e8Trythiww1FWrNfPbukYBmZEbJ",
        ip4: vec!["/ip4/127.0.0.1/tcp/7777"],
        ip6: vec!["/ip6/::1/tcp/9091"]
    },
    AddrInfo {
        id: "12D3KooWGK4RzvNeioS9aXdzmYXU3mgDrRPjQd8SVyXCkHNxLbWN",
        ip4: vec!["/ip4/127.0.0.1/tcp/8080"],
        ip6: vec!["/ip6/::1/tcp/8081"]
    }
    AddrInfo {
        id: "12D3KooWE3kBRAnn6jxZMdK1JMWx1iHtR1NKzXSRv5HLTmfD9u9c",
        ip4: vec![],
        ip6: vec!["/ip6/::1/tcp/5555"],
    }
]
```

So during this process there is no peer in the sense that P2P, or libp2p for that matter, would benefit from it but rather a unique ID and a bunch addresses. When this info is sent over the channel to P2P, libp2p has no use for these anymore so a creating temporary peer object doesn't make sense and creating the peer that P2P will use in libp2p would leak information to the network service provider which doesn't really concern in them in any way.

Reply:
There's probably a misunderstanding. All I'm saying is, everything you're doing inside the loop should move into a separate function, and then you should map that function on the input vector and then collect into the result.

I may have more comments after doing this step. But from what it looks like now, it's very easy call a map instead of a mutable hashmap.

Reply:
Yes I think there is a misunderstanding. I looked at the code and maybe I'm missing some Rust knowledge but what you're suggesting is not as trivial as it sounds and I'm not convinced the end result is cleaner or more maintainable than what the code currently is.

Could you show me how you would do it?

Reply:
If it now looks OK, please approve and I'll squash the commit with the previous commit.

Reply:
Looks perfect now!